### PR TITLE
Do not misuse symex dereferencing code for finding array objects [blocks: #3725]

### PR DESCRIPTION
**Pull request by tautschnig**
The task of process_array_expr is to find the objects pointed to; symex
dereferencing does so as well, but also does many other things. Creating an
artificial dereference_exprt imposes restrictions on what optimisations symex
dereferencing can implement. Instead, just make process_array_expr actually get
the objects (via the value set) and then work on those. This duplicates a bit of
code from symex dereferencing, but now yields full control to process_array_expr
over what it needs to make happen. In future, the process_array_expr code may
even get rewritten without having any dependency on what symex dereferencing
exactly does.

At present, the behaviour before or after the commit is largely unchanged, but modifications such as #3725 show that these changes are actually needed.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: c2dbf9851a5b917d358c56108295deab823411c5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98664986
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/goto-symex/symex_other.cpp [hunk 1]
```diff
@@ -128,7 +131,7 @@ void goto_symext::symex_other(
           statement==ID_array_replace)
   {
     // array_copy and array_replace take two pointers (to arrays); we need to:
-    // 1. dereference the pointers (via clean_expr)
+    // 1. dereference pointers within the pointers (via clean_expr)
```
Comment (smowton) [inline src/goto-symex/symex_other.cpp]: How about just `remove any dereference expressions`, since we're no longer creating a top-level special one?

**Review by smowton — APPROVED** [review]
Looks good except the comment could be clearer

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 36a6215fc644a1b71c18a045b340fd15bf0721fa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98704493

File: src/goto-symex/symex_other.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-symex/symex_other.cpp]: Thanks, done!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: afd28e94a4cce7ba530dd617d9c7caa339dc14d6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98716727

File: src/goto-symex/symex_other.cpp [hunk 2]
```diff
@@ -9,6 +9,9 @@ Author: Daniel Kroening, kroening@kroening.com
 /// \file
 /// Symbolic Execution
 
+#include <iostream>
```
Comment (peterschrammel) [inline src/goto-symex/symex_other.cpp]: 🚫 This shouldn't be needed here.

**Review by peterschrammel — APPROVED** [review]

File: src/goto-symex/symex_other.cpp (hunk 2 above)
Comment (tautschnig) [reply src/goto-symex/symex_other.cpp]: Oops, sorry.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 818b240a01ce36f8b307099d2d496331b22f31fa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98935492
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 855dcfaad693edac589891d7878a02ea075befff).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99078209

**Pull request merged**